### DISCUSSION TITLE
merge pngs into pdf looses quality how too improve my command

### DISCUSSION THREAD

**Discussion by SabineMa** [ImageMagick/ImageMagick#2944] [Help] [answered]
Hi,

here: https://github.com/ImageMagick/ImageMagick/discussions/2784

@fmw42 helped me a lot with his hint for resizing and centering.
But after using that for a while, I found out that I need a better quality of the resulting pdf.

In the following example, the resulting pdf is a lot worser than the original. I ask kindly for help here because I am not an ImageMagick expert at all and I assume that this is more a basic issue....

-> How can I improve my command below for a better quality of the pdf?

`"/usr/local/Cellar/imagemagick/7.0.10-35/bin/magick"  \( "/Users/sabine/Desktop/pdftest/1.png"  \)  \( "/Users/sabine/Desktop/pdftest/2.png"  \)  -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" /Users/sabine/Desktop/pdftest/result.pdf`

For better overview of the problem here is a screenshot of the source png and the result pdf:
![Bildschirmfoto 2020-12-08 um 19 32 03](https://user-images.githubusercontent.com/13763544/101525795-0e1e3f00-398c-11eb-8567-27b6204ce474.png)

This are the  source files (2 times the same just for this example): 
![1](https://user-images.githubusercontent.com/13763544/101525891-36a63900-398c-11eb-8ab5-adb2a2616b51.png)
![2](https://user-images.githubusercontent.com/13763544/101525893-37d76600-398c-11eb-8b01-6ec0c8eacc71.png)

**Comment by snibgo** [comment]
Your inputs are 1190x1684 pixels. The "-resize "%[papersize:A4]>"" operation reduces this to 595x842, which is exactly half the size. So this reduces quality.

I'm not sure what %[papersize:A4] does or is supposed to do. But halving the size, without any change to the aspect ratio, seems like a bad idea. Perhaps it is a bug.

**Comment by Jmuccigr** [comment]
You seem to be on a Mac, so why not just print to pdf and script it via AppleScript if you need to?

Alternatively use img2pdf:

`img2pdf --pagesize A4 -o output.pdf input.jpg`

This will embed the image directly into the pdf.

**Reply by SabineMa** [reply]
@Jmuccigr - Thank you for answering. I can not do this because our application runs on win/unix and mac. And I want to avoid having one more external application to maintain.

**Reply by Jmuccigr** [reply]
OK, I get that adding another dependency can be a hassle, but img2pdf is designed to do _exactly_ what you want and exists for Mac (and other unixen) and Windows (not quite sure what win/unix is). Seems like it might be worth it.

**Comment by fmw42** [comment]
`I'm not sure what %[papersize:A4] does or is supposed to do. But halving the size, without any change to the aspect ratio, seems like a bad idea. Perhaps it is a bug.`


It is rather new and is just a look up to convert between named papersizes and their actual pixel dimension.  It is a convenience function.

**Comment by snibgo** [comment]
Does that help, @SabineMa?

The "-resize" reduces quality by shrinking the images, so don't do it, or the extent.

The perentheses are also redundant, so you can also remove them.

**Comment by SabineMa** [comment]
Hi @snibgo and @fmw42 

Thank you for your reaction.

I have to explain more. I need the resize and extend for that the following requirements are ok:
1) small images should be centered (with the option -gravity center)
2) larger than A4 images should be resized for fitting on the page (note the ">" I learned this from https://github.com/ImageMagick/ImageMagick/discussions/2784)
3) so, each page of the result has to be in A4 format independent of the source size
See https://github.com/ImageMagick/ImageMagick/discussions/2784 for this

I provide a more detailed example with 3 files, one which is small and has to be centered, one which is too large and has to be resized and the one where I have the problem - loosing quality.

with this command: 
`"/usr/local/Cellar/imagemagick/7.0.10-35/bin/magick" "/Users/sabine/Desktop/pdftest/small_center_this.png"  "/Users/sabine/Desktop/pdftest/this_looses_quality.png"  "/Users/sabine/Desktop/pdftest/larger_resize.png"  -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" /Users/sabine/Desktop/pdftest/result.pdf
`

reducing size of parrot is ok, centering is ok but the quality of the text file is not ok.
This is my current Problem and topic of this issue.

![Bildschirmfoto 2020-12-09 um 09 34 03](https://user-images.githubusercontent.com/13763544/101604779-b02c3e80-3a01-11eb-9eab-5773533cbd85.png)

you told me to remove resize and extend.
with this command (without resize and extend): 
`"/usr/local/Cellar/imagemagick/7.0.10-35/bin/magick" "/Users/sabine/Desktop/pdftest/small_center_this.png"  "/Users/sabine/Desktop/pdftest/this_looses_quality.png"  "/Users/sabine/Desktop/pdftest/larger_resize.png"   -gravity center -background white  /Users/sabine/Desktop/pdftest/result.pdf`

The small image grows to page size and I don't have A4 format

![Bildschirmfoto 2020-12-09 um 09 37 21](https://user-images.githubusercontent.com/13763544/101605161-2630a580-3a02-11eb-970f-559912892739.png)

This are the 3 example files:
[Archiv.zip](https://github.com/ImageMagick/ImageMagick/files/5664552/Archiv.zip)

Would be great if you would have a solution for this for me

**Comment by snibgo** [comment, marked as answer]
I don't understand why you want to shrink images. Shrinking images will reduce quality, of course.

Perhaps you want extent, without resizing.

Your three images are 5000x5000, 172x178 and 1190x1684.

We could make them all 5000x7071. This is large enough that none needs resizing, and it is A4 format. So we need only "-extent". For clarity, I use a non-white background. Windows BAT syntax.
```
magick ^
  larger_resize.png ^
  small_center_this.png ^
  this_looses_quality.png ^
  -gravity center -background #45f  ^
  -extent 5000x7071 ^
  out.pdf
```
If you wanted, you could resize all image _up_ to 5000x7071.

**Reply by SabineMa** [reply]
Thank you, @snibgo this is the solution I can use! Thank you for your patience and time!

**Comment by Jmuccigr** [comment]
As @snibgo noted, you can get rid of a lot of those parentheses. On a Mac (with homebrew, it looks like), this:

`"/usr/local/Cellar/imagemagick/7.0.10-35/bin/magick"`

should be able to be just

`magick`

If that doesn't work, something is goofy with your homebrew install. At the very least, homebrew should set up an alias for you at 

`/usr/local/bin/magick`

so you don't need to hard-code the version in the path.

**Reply by SabineMa** [reply]
thanks for the hint. We run imagemagick on different platforms (unix windows and mac) but this should be useful anyway!

**Reply by Jmuccigr** [reply]
Happy to help.

PS I still don't understand the "unix windows and mac" bit. MacOS **is** a flavor of unix, but windows isn't. In any case, I think that any good install of IM should give you a simple "magick" command that will work even when you upgrade, regardless of your OS. If you're using homebrew for Mac, as it appears, you might want to consider their [implementation for linux](https://docs.brew.sh/Homebrew-on-Linux), if that's the unix you're using.

**Reply by fmw42** [reply]
`@Jmuccigr`

Windows 10 now comes with a Unix subsystem. You can open a bash terminal or keep your DOS CMD Window. Unix ImageMagick can be installed in the Windows 10 Unix subsystem and run just like on Linux or Mac OSX. See for example http://www.howtogeek.com/249966/how-to-install-and-use-the-linux-bash-shell-on-windows-10/

**Reply by Jmuccigr** [reply]
Good to know! Thanks.
